File: Level_2/괄호_변환.js
function solution(p) {
  if (p.length === 0) return '';

  let cnt = 0;
  let u = '';
  let v = '';

  // cnt === 0 일 때 문자열 나누기
  // ( = ++ , ) = --
  for (let i = 0; i < p.length; i++) {
    if (p[i] === '(') cnt++;
    else cnt--;

    if (cnt === 0) {
      u = p.slice(0, i + 1);
      v = p.slice(i + 1);

      if (u.indexOf('(') > u.indexOf(')')) {
        let temp = '(';
        temp += solution(v);
        temp += ')';

        u = [...u.slice(1, u.length - 1)];
        for (let i = 0; i < u.length; i++) {
          if (u[i] === '(') u[i] = ')';
          else u[i] = '(';
        }
        u = u.join('');

        temp += u;

        return temp;
      }

      return u + solution(v);
    }
  }
}

let p = '(()())()';
let p_1 = '()))((()';
let p_2 = ')(';

console.log(solution(p));
console.log(solution(p_1));
console.log(solution(p_2));

File: Level_2/괄호_변환.py
def solution(p):
    answer = ''
    balance = 0
    correct = True
    u = ''
    v = ''

    if len(p) == 0:
        return ''

    for i in range(len(p)):
        if p[i] == '(':
            balance += 1
        else:
            balance -= 1

        if i == 0 and balance < 0:
            correct = False

        if balance == 0:
            u = p[:i+1]
            v = p[i+1:]
            break

    if not correct:
        temp = '('
        temp += solution(v)
        temp += ')'
        u = list(u[1:len(u)-1])
        for i in range(len(u)):
            if u[i] == '(':
                u[i] = ')'
            elif u[i] == ')':
                u[i] = '('
        u = ''.join(u)
        temp += u
        return temp

    answer = u + solution(v)

    return answer


p = '(()())()'
print(solution(p))
p = ')('
print(solution(p))
p = '()))((()'
print(solution(p))
